packages/@lwc/engine-core/src/framework/stylesheet.ts

```diff
@@ -138,16 +138,39 @@ export function getStylesheetsContent(vm: VM, template: Template): string[] {
     return content;
 }
 
+function getNearestNativeShadowComponent(vm: VM): VM | null {
+    let owner: VM | null = vm;
+    while (!isNull(owner)) {
+        if (owner.renderMode === RenderMode.Shadow && owner.shadowMode === ShadowMode.Native) {
+            return owner;
+        }
+        owner = owner.owner;
+    }
+    return owner;
+}
```

Comment:
Would it be better performance-wise to store this on the VM instead of looking this up lazily?

Reply:
This is worth testing, although we'd have to be careful about when to invalidate the cache (e.g. if the component is moved).

Reply:
@pmdartus I tried it (https://github.com/nolanlawson/lwc-1/commit/87345eeaf02b8778bc4776ad9e575414347bb4b8), but it doesn't seem to be a statistically-significant perf improvement, at least in these benchmarks. I also worry that it introduces footguns (e.g. if we don't invalidate the cache at the right time).

There might be some benchmark that can prove an improvement (e.g. rendering many stylesheets on deeply-nested components inside a native shadow root), but I don't know how realistic this is.

<details><summary>Benchmark results</summary>

```
⠋ Auto-sample 550 (timeout in 0m0s)
┌─────────────┬─────────────────┐
│     Browser │ chrome-headless │
│             │ 92.0.4515.159   │
├─────────────┼─────────────────┤
│ Sample size │ 600             │
└─────────────┴─────────────────┘

┌─────────────────────────────────────────────────┬─────────────┬────────────┬───────────────────┬────────────────────────────────────────────────────┬────────────────────────────────────────────────────┐
│ Benchmark                                       │ Version     │ Bytes      │          Avg time │ vs dom-styled-component-create-1k-same-this-change │ vs dom-styled-component-create-1k-same-tip-of-tree │
│                                                 │             │            │                   │                                                    │                                        tip-of-tree │
├─────────────────────────────────────────────────┼─────────────┼────────────┼───────────────────┼────────────────────────────────────────────────────┼────────────────────────────────────────────────────┤
│ dom-styled-component-create-1k-same-this-change │ <none>      │ 220.48 KiB │ 52.77ms - 53.43ms │                                                    │                                             unsure │
│                                                 │             │            │                   │                                           -        │                                          -0% - +1% │
│                                                 │             │            │                   │                                                    │                                  -0.19ms - +0.77ms │
├─────────────────────────────────────────────────┼─────────────┼────────────┼───────────────────┼────────────────────────────────────────────────────┼────────────────────────────────────────────────────┤
│ dom-styled-component-create-1k-same-tip-of-tree │ tip-of-tree │ 220.18 KiB │ 52.46ms - 53.16ms │                                             unsure │                                                    │
│                                                 │             │            │                   │                                          -1% - +0% │                                           -        │
│                                                 │             │            │                   │                                  -0.77ms - +0.19ms │                                                    │
└─────────────────────────────────────────────────┴─────────────┴────────────┴───────────────────┴────────────────────────────────────────────────────┴────────────────────────────────────────────────────┘

NOTE Hit 5 minute auto-sample timeout trying to resolve horizon(s)

⠋ Auto-sample 330 (timeout in 0m44s)

┌─────────────┬─────────────────┐
│     Browser │ chrome-headless │
│             │ 92.0.4515.159   │
├─────────────┼─────────────────┤
│ Sample size │ 380             │
└─────────────┴─────────────────┘

┌──────────────────────────────────────────────────────┬─────────────┬────────────┬─────────────────────┬─────────────────────────────────────────────────────────┬─────────────────────────────────────────────────────────┐
│ Benchmark                                            │ Version     │ Bytes      │            Avg time │ vs dom-styled-component-create-1k-different-this-change │ vs dom-styled-component-create-1k-different-tip-of-tree │
│                                                      │             │            │                     │                                                         │                                             tip-of-tree │
├──────────────────────────────────────────────────────┼─────────────┼────────────┼─────────────────────┼─────────────────────────────────────────────────────────┼─────────────────────────────────────────────────────────┤
│ dom-styled-component-create-1k-different-this-change │ <none>      │ 977.19 KiB │ 157.52ms - 159.63ms │                                                         │                                                  unsure │
│                                                      │             │            │                     │                                                -        │                                               -1% - +1% │
│                                                      │             │            │                     │                                                         │                                       -1.20ms - +1.57ms │
├──────────────────────────────────────────────────────┼─────────────┼────────────┼─────────────────────┼─────────────────────────────────────────────────────────┼─────────────────────────────────────────────────────────┤
│ dom-styled-component-create-1k-different-tip-of-tree │ tip-of-tree │ 976.89 KiB │ 157.50ms - 159.29ms │                                                  unsure │                                                         │
│                                                      │             │            │                     │                                               -1% - +1% │                                                -        │
│                                                      │             │            │                     │                                       -1.57ms - +1.20ms │                                                         │
└──────────────────────────────────────────────────────┴─────────────┴────────────┴─────────────────────┴─────────────────────────────────────────────────────────┴─────────────────────────────────────────────────────────┘

```

(Note tip-of-tree is the `nolan/constructable-stylesheets-redux` branch in this case.)

</details>